python/ray/_common/network_utils.py

```diff
@@ -29,6 +35,65 @@ def build_address(host: str, port: Union[int, str]) -> str:
     return _build_address(host, port)
 
 
+def node_ip_address_from_perspective(address: str = "") -> str:
+    """IP address by which the local node can be reached *from* the `address`.
+
+    If no address is given, defaults to public DNS servers for detection. For
+    performance, the result is cached when using the default address (empty string).
+    When a specific address is provided, detection is performed fresh every time.
+
+    Args:
+        address: The IP address and port of any known live service on the
+            network you care about.
+
+    Returns:
+        The IP address by which the local node can be reached from the address.
+    """
+    return _node_ip_address_from_perspective(address)
+
+
+def is_ipv6_ip(ip: str) -> bool:
+    """Check if an IP string is IPv6 format.
+
+    Args:
+        ip: The IP address string to check (must be pure IP, no port).
+
+    Returns:
+        True if the IP is IPv6, False if IPv4.
+    """
+    return _is_ipv6_ip(ip)
+
+
+@lru_cache(maxsize=1)
+def get_localhost_address() -> str:
+    """Get localhost loopback address with IPv4/IPv6 support.
+
+    Returns:
+        The localhost loopback IP address (matching node IP family or auto-detected).
+    """
+    import ray._private.worker
+
+    if (
+        ray._private.worker._global_node is not None
+        and ray._private.worker._global_node.node_ip_address
+    ):
+        node_ip = ray._private.worker._global_node.node_ip_address
+        return "::1" if is_ipv6_ip(node_ip) else "127.0.0.1"
```

Comment:
I don't think we need to check node_ip for localhost: it's ok that public ip is ipv6 and localhost is 127.0.0.1